src/algebra/square.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2022 Kevin H. Wilson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin H. Wilson
+-/
+import ring_theory.int.basic
+
+/-!
+# Square elements of monoids
+An element `x` of a monoid is a square when there is another element `y` such that `y * y = x`.
+
+## Main Definitions
+ - `square r` indicates that `∃ s, s * s = r`.
+
+## Main Results
+ - `nat.square_iff_factorization_even`: `x` is a `square` iff every entry in its factorization
+   is even
+ - `nat.square_mul`: Being `square` is multiplicative
+
+## TODO
+
+There are many more facts about squares that generalize to arbitrary `unique_factorization_monoid`s.
+The should eventually be added.
+
+## Tags
+square, squarefree, multiplicity
+
+-/
+
+variables {R : Type*}
+
+/-- An element of a monoid is squarefree if the only squares that
+  divide it are the squares of units. -/
+def square [monoid R] (r : R) : Prop := ∃ x : R, x * x = r
+
+@[simp]
+lemma square_one [comm_monoid R] : square (1 : R) := by { use 1, simp, }
+
+@[simp]
+lemma square_zero [monoid_with_zero R] : square (0 : R) :=
+by { use 0, simp, }
+
+@[simp]
+lemma irreducible.not_square [comm_monoid R] {x : R} (h : irreducible x) :
+  ¬square x :=
+begin
+  rintros ⟨y, hy⟩,
+  rcases h.is_unit_or_is_unit hy.symm with hu | hu,
+  { rw ← hy at h,
+    exact h.not_unit (hu.mul hu), },
+  { rw ← hy at h,
+    exact h.not_unit (hu.mul hu), },
+end
+
+@[simp]
+lemma prime.not_square [cancel_comm_monoid_with_zero R] {x : R} (h : prime x) :
+  ¬square x := h.irreducible.not_square
+
+/-- The product of squares is a square. For a partial converse, see `nat.square_mul` -/
+lemma square_mul_of_square_of_square {R : Type*} [comm_monoid R] {m n : R} :
+  square m → square n → square (m * n) :=
+begin
+  rintros ⟨c, hc⟩,
+  rintros ⟨d, hd⟩,
+  use (c * d),
+  have : c * d * (c * d) = (c * c) * (d * d),
+  { rw mul_assoc,
+    conv in (d * (c * d)) { rw [mul_comm, mul_assoc], },
+    rw ←mul_assoc, },
+  rw [this, hc, hd],
+end
+
+section factorization
+
+namespace nat
+
+instance : decidable_pred (square : ℕ → Prop)
+| n := decidable_of_iff' _ (nat.exists_mul_self n)
+
+lemma square_iff_factorization_even {m : ℕ} :
+square m ↔ ∀ (p : ℕ), even (m.factorization p) :=
+begin
+  by_cases hm_zero : m = 0,
+  { simp only [hm_zero, square_zero, factorization_zero, finsupp.coe_zero,
+      pi.zero_apply, forall_const, true_iff],
+    use 0,
```

Comment:
The remaining goal here is `even 0`, surely this could be its own `simp` lemma? If not, you can save a line by merging this and the following line into
```lean
exact ⟨0, mul_zero 2⟩
```
Same goes for another usage of this a few lines down.